### PR TITLE
Change D_METHOD to variadic template

### PR DESCRIPTION
**Pull request by KoBeWi**
Copy-pasted from UndoRedo code.

### CODE REVIEW

**Review by akien-mga — APPROVED** [review]

**Comment by KoBeWi** [comment]
Rebased after #59932
Release build still compiles, so I guess it's alright.

File: core/object/class_db.cpp [hunk 1]
```diff
@@ -39,190 +39,13 @@
...
-	md.args.write[5] = StaticCString::create(p_arg6);
-	md.args.write[6] = StaticCString::create(p_arg7);
-	md.args.write[7] = StaticCString::create(p_arg8);
-	md.args.write[8] = StaticCString::create(p_arg9);
-	md.args.write[9] = StaticCString::create(p_arg10);
-	md.args.write[10] = StaticCString::create(p_arg11);
-	md.args.write[11] = StaticCString::create(p_arg12);
-	md.args.write[12] = StaticCString::create(p_arg13);
+	md.args.resize(p_argcount);
+	for (uint64_t i = 0; i < p_argcount; i++) {
```
Comment (akien-mga) [inline core/object/class_db.cpp:46]: core/object/class_db.cpp: In function 'MethodDefinition D_METHODP(const char*, const char**, int)':
core/object/class_db.cpp:46:25: error: comparison of integer expressions of different signedness: 'uint64_t' {aka 'long unsigned int'} and 'int' [-Werror=sign-compare]
   46 |  for (uint64_t i = 0; i < p_argcount; i++) {
      |                       ~~^~~~~~~~~~~~

Comment (akien-mga) [inline core/object/class_db.cpp:46]: core/object/class_db.cpp: In function 'MethodDefinition D_METHODP(const char*, const char**, int)':
core/object/class_db.cpp:46:25: error: comparison of integer expressions of different signedness: 'uint64_t' {aka 'long unsigned int'} and 'int' [-Werror=sign-compare]
   46 |  for (uint64_t i = 0; i < p_argcount; i++) {
      |                       ~~^~~~~~~~~~~~

Comment (KoBeWi) [reply core/object/class_db.cpp:46]: Looks like code in UndoRedo is wrong. It uses `uint` in a for loop and then passes the arguments to method that takes `int`.

**Comment by akien-mga** [comment]
It seems to make the Mono build fail. Any clue @godotengine/mono?
```
/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/AESContext.cs(125,26): warning CS1658: Unexpected character '�'. See also error CS1056. [/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/AESContext.cs(125,27): warning CS1658: Unexpected character '�'. See also error CS1056. [/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/AESContext.cs(125,29): warning CS1658: Unexpected character '�'. See also error CS1056. [/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/AESContext.cs(125,30): warning CS1658: Unexpected character '�'. See also error CS1056. [/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/AESContext.cs(125,31): warning CS1658: Unexpected character ''. See also error CS1056. [/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/AESContext.cs(127,44): error CS1001: Identifier expected [/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
```

**Comment by neikeq** [comment]
I can't test this PR right now, but my guess is that it messes up the argument names. I'd look at the `AESContext` docs in the editor to confirm.

**Comment by akien-mga** [comment]
> I can't test this PR right now, but my guess is that it messes up the argument names. I'd look at the `AESContext` docs in the editor to confirm.

Indeed, I compiled the PR and it has some issues.

Running the editor spams this error:
```
ERROR: Condition "!p_static_string.ptr || !p_static_string.ptr[0]" is true.
   at: StringName (core/string/string_name.cpp:269)
```

And indeed argument names are messed up:
image

**Comment by akien-mga** [comment]
I don't know what I'm doing but this seems to work for me:
```diff
diff --git a/core/object/class_db.cpp b/core/object/class_db.cpp
index 290330d718..7c411263df 100644
--- a/core/object/class_db.cpp
+++ b/core/object/class_db.cpp
@@ -39,12 +39,12 @@
 
 #ifdef DEBUG_METHODS_ENABLED
 
-MethodDefinition D_METHODP(const char *p_name, const char **p_args, uint64_t p_argcount) {
+MethodDefinition D_METHODP(const char *p_name, const char ***p_args, uint64_t p_argcount) {
 	MethodDefinition md;
 	md.name = StaticCString::create(p_name);
 	md.args.resize(p_argcount);
 	for (uint64_t i = 0; i < p_argcount; i++) {
-		md.args.write[i] = StaticCString::create(p_args[i]);
+		md.args.write[i] = StaticCString::create(*p_args[i]);
 	}
 	return md;
 }
diff --git a/core/object/class_db.h b/core/object/class_db.h
index 83aeef2d01..31cf4e94dc 100644
--- a/core/object/class_db.h
+++ b/core/object/class_db.h
@@ -35,10 +35,6 @@
 #include "core/object/object.h"
 #include "core/string/print_string.h"
 
-/** To bind more then 6 parameters include this:
- *
- */
-
 // Makes callable_mp readily available in all classes connecting signals.
 // Needs to come after method_bind and object have been included.
 #include "core/object/callable_method_pointer.h"
@@ -57,7 +53,7 @@ struct MethodDefinition {
 			name(p_name) {}
 };
 
-MethodDefinition D_METHODP(const char *p_name, const char **p_args, uint64_t p_argcount);
+MethodDefinition D_METHODP(const char *p_name, const char ***p_args, uint64_t p_argcount);
 
 template <typename... VarArgs>
 MethodDefinition D_METHOD(const char *p_name, const VarArgs... p_args) {
@@ -67,7 +63,7 @@ MethodDefinition D_METHOD(const char *p_name, const VarArgs... p_args) {
 		argptrs[i] = &args[i];
 	}
 
-	return D_METHODP(p_name, sizeof...(p_args) == 0 ? nullptr : (const char **)argptrs, sizeof...(p_args));
+	return D_METHODP(p_name, sizeof...(p_args) == 0 ? nullptr : (const char ***)argptrs, sizeof...(p_args));
 }
 
 #else
```

But I'd get an expert C/C++ dev to check this `const char ***` business as I'm really not sure that the `const` applies as it's meant to.

**Comment by akien-mga** [comment]
New diff suggestion which I think preserves the pointer constness intended in other similar variadic templates (and changes `uint64_t` to `uint32_t` as suggested):
```diff
diff --git a/core/object/class_db.cpp b/core/object/class_db.cpp
index 290330d718..b561765495 100644
--- a/core/object/class_db.cpp
+++ b/core/object/class_db.cpp
@@ -39,12 +39,12 @@
 
 #ifdef DEBUG_METHODS_ENABLED
 
-MethodDefinition D_METHODP(const char *p_name, const char **p_args, uint64_t p_argcount) {
+MethodDefinition D_METHODP(const char *p_name, const char * const **p_args, uint32_t p_argcount) {
 	MethodDefinition md;
 	md.name = StaticCString::create(p_name);
 	md.args.resize(p_argcount);
 	for (uint64_t i = 0; i < p_argcount; i++) {
-		md.args.write[i] = StaticCString::create(p_args[i]);
+		md.args.write[i] = StaticCString::create(*p_args[i]);
 	}
 	return md;
 }
diff --git a/core/object/class_db.h b/core/object/class_db.h
index 83aeef2d01..b22f109423 100644
--- a/core/object/class_db.h
+++ b/core/object/class_db.h
@@ -35,10 +35,6 @@
 #include "core/object/object.h"
 #include "core/string/print_string.h"
 
-/** To bind more then 6 parameters include this:
- *
- */
-
 // Makes callable_mp readily available in all classes connecting signals.
 // Needs to come after method_bind and object have been included.
 #include "core/object/callable_method_pointer.h"
@@ -57,17 +53,17 @@ struct MethodDefinition {
 			name(p_name) {}
 };
 
-MethodDefinition D_METHODP(const char *p_name, const char **p_args, uint64_t p_argcount);
+MethodDefinition D_METHODP(const char *p_name, const char * const **p_args, uint32_t p_argcount);
 
 template <typename... VarArgs>
 MethodDefinition D_METHOD(const char *p_name, const VarArgs... p_args) {
 	const char *args[sizeof...(p_args) + 1] = { p_args..., nullptr }; // +1 makes sure zero sized arrays are also supported.
-	const char **argptrs[sizeof...(p_args) + 1];
+	const char * const *argptrs[sizeof...(p_args) + 1];
 	for (uint32_t i = 0; i < sizeof...(p_args); i++) {
 		argptrs[i] = &args[i];
 	}
 
-	return D_METHODP(p_name, sizeof...(p_args) == 0 ? nullptr : (const char **)argptrs, sizeof...(p_args));
+	return D_METHODP(p_name, sizeof...(p_args) == 0 ? nullptr : (const char * const **)argptrs, sizeof...(p_args));
 }
 
 #else
```

Still compiling to see if it actually works.

**Review by akien-mga — APPROVED** [review]
Seems to build and work fine now :+1:

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**